### PR TITLE
STYLE: fix some consider-using-enumerate pylint warnings

### PR DESCRIPTION
**Pull request by Moisan**
Related to https://github.com/pandas-dev/pandas/issues/48855
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

This covers some `consider-using-enumerate` pylint warnings but not all them. The ones left are often using the index of the for-loop without accessing element of the list. Is that something we want to put an "ignore" tag on?

### CODE REVIEW

File: pandas/core/array_algos/take.py [hunk 1]
```diff
@@ -549,13 +549,11 @@ def _take_2d_multi_object(
             out[row_mask, :] = fill_value
         if col_needs:
             out[:, col_mask] = fill_value
-    for i in range(len(row_idx)):
+    for i, u_ in enumerate(row_idx):
         u_ = row_idx[i]
```
Comment (mroeschke) [inline pandas/core/array_algos/take.py:553]: This `u_` shouldn' be needed anymore I think

**Comment by mroeschke** [comment]
> The ones left are often using the index of the for-loop without accessing element of the list. Is that something we want to put an "ignore" tag on?

These should be refactored if possible

File: pandas/core/reshape/reshape.py [hunk 2]
```diff
@@ -634,20 +635,13 @@ def stack_multiple(frame, level, dropna: bool = True):
         # negative numbers to positive
         level = [frame.columns._get_level_number(lev) for lev in level]
 
-        # Can't iterate directly through level as we might need to change
-        # values as we go
-        for index in range(len(level)):
-            lev = level[index]
+        level.reverse()
+        while level:
+            lev = level.pop()
```
Comment (MarcoGorelli) [inline pandas/core/reshape/reshape.py:640]: would `lev = level.pop(0)` (without `level.reverse`) work?

**Review by mroeschke — APPROVED** [review]
LGTM cc @MarcoGorelli merge when ready

File: pandas/core/reshape/reshape.py [hunk 3]
```diff
@@ -634,20 +634,12 @@ def stack_multiple(frame, level, dropna: bool = True):
...
             # Decrement all level numbers greater than current, as these
             # have now shifted down by one
-            updated_level = []
-            for other in level:
-                if other > lev:
-                    updated_level.append(other - 1)
-                else:
-                    updated_level.append(other)
-            level = updated_level
+            level = [v if v <= lev else v - 1 for v in level]
```
Comment (MarcoGorelli) [inline pandas/core/reshape/reshape.py:642]: Nice! Looks fine, and I ran
```
$ cat f.py 
import numpy as np

for _ in range(10):

    original_level = np.random.randint(0, 8, size=9)

    main_lev_sequence = []
    level = [i for i in original_level]
    for index in range(len(level)):
        lev = level[index]
        main_lev_sequence.append(lev)
        updated_level = []
        for other in level:
            if other > lev:
                updated_level.append(other - 1)
            else:
                updated_level.append(other)
        level = updated_level

    branch_lev_sequence = []
    level = [i for i in original_level]
    while level:
        lev = level.pop(0)
        branch_lev_sequence.append(lev)
        level = [v if v <= lev else v - 1 for v in level]

    assert main_lev_sequence == branch_lev_sequence
```
to check, which worked

**Review by MarcoGorelli — APPROVED** [review]
Thanks @Moisan !

**Pull request merged**